azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+from azurelinuxagent.common.utils.shellutil import run_command
+from azurelinuxagent.common.utils.textutil import safe_shlex_split
+
+
+LOG_COLLECTOR_DIR = '/var/lib/waagent/logcollector'
+TRUNCATED_FILES_DIR = '/var/truncated'
+
+OUTPUT_RESULTS_FILE_PATH = os.path.join(LOG_COLLECTOR_DIR, "results.txt")
+OUTPUT_ARCHIVE_PATH = os.path.join(LOG_COLLECTOR_DIR, "logs.tar")
+COMPRESSED_ARCHIVE_PATH = os.path.join(LOG_COLLECTOR_DIR, "logs.zip")
+
+MUST_COLLECT_FILES = [
+    '/var/log/waagent*',
+    '/var/lib/waagent/GoalState.*.xml',
+    '/var/lib/waagent/waagent_status.json',
+    '/var/lib/waagent/history/*.zip',
+    '/var/log/azure/*/*',
+    '/var/log/azure/*/*/*',
+    '/var/lib/waagent/ExtensionsConfig.*.xml',
+    '/var/lib/waagent/HostingEnvironmentConfig.xml',
+    '/var/lib/waagent/error.json'
+]
+
+FILE_SIZE_LIMIT = 30 * 1024 * 1024  # 30 MB
+ARCHIVE_SIZE_LIMIT = 150 * 1024 * 1024  # 150 MB
+
+
+class LogCollector(object):
+
+    def __init__(self, manifest_file_path):
+        self.manifest_file_path = manifest_file_path
+        self.must_collect_files = self._expand_must_collect_files()
+
+    def _expand_must_collect_files(self):
+        # Match the regexes from the MUST_COLLECT_FILES list to existing file paths on disk.
+        manifest = []
+        for path in MUST_COLLECT_FILES:
+            expanded_paths = self._expand_path(path)
+            manifest.extend(expanded_paths)
+
+        return manifest
+
+    def _read_manifest_file(self):
+        return read_file(self.manifest_file_path).splitlines()
+
+    @staticmethod
+    def _log_to_results_file(entry):
+        if type(entry) == ustr or type(entry) == str:
+            append_file(OUTPUT_RESULTS_FILE_PATH, entry + "\n")
+        else:
+            append_items_to_file(OUTPUT_RESULTS_FILE_PATH, entry)
+
+    @staticmethod
+    def _expand_path(path):
+        # If the path needs expanding, e.g. '/var/log/azure/*', add all file matches, sorted by name.
+        # Otherwise, add the already expanded path, e.g. '/var/lib/waagent/HostingEnvironmentConfig.xml'
+        expanded = []
+        if os.path.exists(path):
+            expanded.append(path)
+        else:
+            paths = glob.glob(path)
+            if len(paths) > 0:
+                expanded.extend(sorted(paths))
+        return expanded
+
+    @staticmethod
+    def _parse_ll_command(folder):
+        output = run_command(["ls", "-alF", folder])
+        header = "Output of \"ll {0}\":\n".format(folder)
+        LogCollector._log_to_results_file(header + output)
+
+    @staticmethod
+    def _parse_echo_command(message):
+        LogCollector._log_to_results_file(message)
+
+    @staticmethod
+    def _parse_copy_command(path):
+        file_paths = LogCollector._expand_path(path)
+        LogCollector._log_to_results_file(file_paths)
+        return file_paths
+
+    @staticmethod
+    def _convert_file_name_to_archive(file_name):
```

Comment:
can you add comments as to what "file name" and "archive name" are? i know what they are, but it is not obvious from the code

Reply:
Added comments.